File: qmc_uniform.py
#!/usr/bin/env python3

from hydrogen  import *
from qmc_stats import *
import numpy as np

def MonteCarlo(a, nmax):
     energy = 0.
     normalization = 0.

     for istep in range(nmax):
          #r = np.random.uniform(-5., 5., (3))
          R = 5.
          phi = np.random.rand()*2*np.pi
          costheta = np.random.rand()*2 - 1.0
          u = np.random.rand()

          theta = np.arccos( costheta )
          r = R * np.cbrt( u )

          # Spherical distribution
          x = r * np.sin( theta) * np.cos( phi )
          y = r * np.sin( theta) * np.sin( phi )
          z = r * np.cos( theta )
          r = np.array([x,y,z],dtype=np.Float64)


          w = psi(a,r)
          w = w*w

          energy        += w * e_loc(a,r)
          normalization += w

     return energy / normalization

a    = 1.2
nmax = 100000

X = [MonteCarlo(a,nmax) for i in range(30)]
E, deltaE = ave_error(X)

print(f"E = {E} +/- {deltaE}")

File: qmc_uniform.c
#include <stdlib.h>
#include <math.h>
#include <stdio.h>
#include <time.h>
#include <stddef.h>  // for size_t
#include "hydrogen.h"
#include "qmc_stats.h"   // for ave_error

void uniform_montecarlo(double a, size_t nmax, double *energy) {
    size_t istep;
    double normalization, r[3], w, f;

    *energy = 0.0;
    normalization = 0.0;

    for (istep = 0; istep < nmax; istep++) {
        for (int i = 0; i < 3; i++) {
            r[i] = drand48();
        }

        r[0] = -5.0 + 10.0 * r[0];
        r[1] = -5.0 + 10.0 * r[1];
        r[2] = -5.0 + 10.0 * r[2];
        f = psi(a, r);
        w = f*f;
        *energy += w * e_loc(a, r);
        normalization += w;
    }
    *energy = *energy / normalization;
}

int main(void) {

#define a     1.2
#define nmax  100000
#define nruns 30

    double X[nruns];
    double ave, err;

    srand48(time(NULL));

    for (size_t irun = 0; irun < nruns; irun++) {
        uniform_montecarlo(a, nmax, &X[irun]);
    }
    ave_error(X, nruns, &ave, &err);

    printf("E = %f +/- %f\n", ave, err);

    return 0;
}
